daemon/cluster/controllers/plugin/controller.go

```diff
@@ -1,32 +1,140 @@
+	Remove(name string, config *enginetypes.PluginRmConfig) error
+	Pull(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer, opts ...plugin.CreateOpt) error
+	Upgrade(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer) error
+	Get(name string) (*v2.Plugin, error)
+	SubscribeEvents(buffer int, events ...plugin.Event) (eventCh <-chan interface{}, cancel func())
+}
 
 // NewController returns a new cluster plugin controller
-func NewController() (*Controller, error) {
-	return &Controller{}, nil
+func NewController(backend Backend, t *api.Task) (*Controller, error) {
+	spec, err := readPayload(t)
+	if err != nil {
+		return nil, err
+	}
+	return &Controller{backend: backend, spec: spec, serviceID: t.ServiceID}, nil
 }
 
 // Update is the update phase from swarmkit
 func (p *Controller) Update(ctx context.Context, t *api.Task) error {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Update")
+
 	return nil
 }
 
+func readPayload(t *api.Task) (runtime.GenericPluginPayload, error) {
+	var cfg runtime.GenericPluginPayload
+
+	generic := t.Spec.GetGeneric()
+	if err := proto.Unmarshal(generic.Payload.Value, &cfg); err != nil {
+		return cfg, errors.Wrap(err, "error reading plugin config")
+	}
+	return cfg, nil
+}
+
 // Prepare is the prepare phase from swarmkit
-func (p *Controller) Prepare(ctx context.Context) error {
+func (p *Controller) Prepare(ctx context.Context) (err error) {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Prepare")
+
+	// TODO(@cpuguy83): support digest pinning
+	remote, err := reference.ParseNormalizedNamed(p.spec.Spec.Remote)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing remote reference %q", p.spec.Spec.Remote)
+	}
+
+	if p.spec.Spec.Name == "" {
+		p.spec.Spec.Name = remote.String()
+	}
+
+	// retrieve auth config from encoded auth
+	var authConfig enginetypes.AuthConfig
+	if p.spec.PullOptions != nil && p.spec.PullOptions.RegistryAuth != "" {
+		if err := json.NewDecoder(base64.NewDecoder(base64.URLEncoding, strings.NewReader(p.spec.PullOptions.RegistryAuth))).Decode(&authConfig); err != nil {
+			logrus.Warnf("invalid authconfig: %v", err)
+		}
+	}
+
+	privs := convertPrivileges(p.spec.Spec.Privileges)
+
+	pl, err := p.backend.Get(p.spec.Spec.Name)
+
+	defer func() {
+		if pl != nil && err == nil {
+			pl.Acquire()
+		}
+	}()
+
+	if err == nil && pl != nil {
+		if pl.SwarmServiceID != p.serviceID {
+			return errors.Errorf("plugin already exists: %s", p.spec.Spec.Name)
+		}
+		if pl.IsEnabled() {
+			if err := p.backend.Disable(pl.GetID(), &enginetypes.PluginDisableConfig{ForceDisable: false}); err != nil {
+				logrus.WithFields(logrus.Fields{
+					"controller": "plugin",
+					"action":     "prepare",
```

Comment:
I'd recommend moving this above so we can use it with the `invalid authconfig` log message above, which otherwise is lacking context. I think these fields should also include the task ID.